src/components/Datepicker/Datepicker.js

```diff
@@ -0,0 +1,159 @@
+import React, { useState } from 'react'
+import PropTypes from 'prop-types'
+import { useDatepicker, START_DATE } from '@datepicker-react/hooks'
+import { noop } from '../../utilities/other'
+import { getJSDateFromString, getValidDateTimeString } from './Datepicker.utils'
+import DatepickerContext from './Datepicker.Context'
+import { CalendarContainerUI } from './Datepicker.css'
+import PeriodCalendar from './Datepicker.PeriodCalendar'
+import DailyCalendar from './Datepicker.DailyCalendar'
+
+function Datepicker({
+  allowFutureDatePick,
+  endDate,
+  firstDayOfWeek,
+  minBookingDays,
+  numberOfMonths,
+  onDateChange,
+  startDate,
+}) {
+  const [dates, setDates] = useState({
+    startDate: getJSDateFromString(startDate),
+    endDate: getJSDateFromString(endDate) || getJSDateFromString(startDate),
+    focusedInput: START_DATE,
+  })
+  const [prevDates, setPrevDates] = useState({})
+  const [deepNavVisible, setDeepNavVisibility] = useState(false)
+
+  const {
+    activeMonths,
+    isDateSelected,
+    isDateHovered,
+    isFirstOrLastSelectedDate,
+    isDateBlocked,
+    isDateFocused,
+    onDateSelect,
+    goToDate,
+    goToPreviousMonthsByOneMonth,
+    goToNextMonthsByOneMonth,
+    goToPreviousYear,
+    goToNextYear,
+  } = useDatepicker({
+    startDate: dates.startDate,
+    endDate: dates.endDate,
+    focusedInput: dates.focusedInput,
+    onDatesChange: handleDateChange,
+    numberOfMonths,
+    exactMinBookingDays: minBookingDays === 1,
+    maxBookingDate: !allowFutureDatePick ? new Date() : undefined,
+  })
+
+  /**
+   * This section performs a `getDerivedStateFromProps` style logic where we
+   * store the prop in state (as prevDate) so we can compare in the next render
+   * and only if the prop changes this gets run again
+   */
+  if (
```

Comment:
useEffect is here for you, for exactly that reason. It will execute code based on a variable, when those change

you might have tried it, and determine that it was not doing what you need, and that's fine.

a possibility would be to separate the state for each key:

```javascript
const [startDate, setStartDate] = useState(getJSDateFromString(startDateProp))
const [endDate, setEndDate] = useState(getJSDateFromString(endDateProp) || startDate)
```
qq: Is there a moment where focusedInput could change? if not you might just use the constant in useDatepicker. 

then useEffect for updating the date
```javascript
useEffect( () => {
  setStartDate(getJSDateFromString(startDateProp))
  setEndDate(getJSDateFromString(endDateProp))
}, [getValidDateTimeString (startDateProp), getValidDateTimeString (endDateProp)])
```

and finally in handleDateChange you could set each date state:

```javascript
function handleDateChange(data) {
    setStartDate(data.startDate)
    setEndDate(data.endDate)
    onDateChange(data)
  }
```

I did not test that, it could be different dependencies for useEffect. You know the code better than me, so I will let you judge if that update would be beneficial in the end 🎉

Reply:
Using `useEffect` is how I first implemented this part, and it worked ok-ish: I didn't like the fact that it would always trigger a re-render on the component initial mounting.

There are 2 things that need to happen when the prop changes: 
1. Select the date (this is what the `setDates` call is responsible for)
2. Navigate the calendar to that date (done by the function `goToDate` provided by datepicker-hooks, I believe this uses setState under the hood to set some "navigation" state)

Number 2 is an unnecessary call most of the time except for this particular case (it gets called for you in most scenarios, like navigating month by month), without it the new date would be selected but you would have to navigate manually to the correct year/month to see it.

So what ends up happening with the use of `useEffect` is that you set the dates using `setState` and then after the render (or mounting) happens, it runs the `setState` and `goToDate`, triggering a re-render.

The code here ensures not having to do something more complicated just to be able to use `useEffect` (given that `goToDate` is a dependency for example, which means something weird like having to use `useCallback` for it, or having to add extra conditionals to check the "navigation" state of the calendar, etc), and it performs as expected with no re-renders and no effect other than if a different date is passed as a prop, and my guess is that changing the date via prop in something other than the initial render rarely happens.

Hope that clarifies things a bit, thanks for the notes!